planner/core/find_best_task.go

```diff
@@ -582,6 +582,18 @@ func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candida
 			candidates = append(candidates, currentCandidate)
 		}
 	}
+
+	if ds.ctx.GetSessionVars().GetAllowAlwaysPreferIndex() {
+		if len(candidates) > 1 {
+			for i, c := range candidates {
+				if c.path.IsTablePath() {
+					candidates = append(candidates[:i], candidates[i+1:]...)
```

Comment:
What if the table path is tableRangeScan and the range is very small and index range is very large?

Reply:
You can refer to the issue #15623 about this problem.

Reply:
I think it is better to exclude only full table scan rather than exclude all table scan.

Reply:
@coocood The code has been modified, please check if it is correct. ^_^